srml/staking/src/phragmen.rs

```diff
@@ -0,0 +1,223 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Rust implementation of the Phragmén election algorithm.
+
+use rstd::{prelude::*};
+use primitives::Perquintill;
+use primitives::traits::{Zero, As};
+use parity_codec::{HasCompact, Encode, Decode};
+use crate::{Exposure, BalanceOf, Trait, ValidatorPrefs, IndividualExposure};
+
+// Wrapper around validation candidates some metadata.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Candidate<AccountId, Balance: HasCompact> {
+	// The validator's account
+	pub who: AccountId,
+	// Exposure struct, holding info about the value that the validator has in stake.
+	pub exposure: Exposure<AccountId, Balance>,
+	// Accumulator of the stake of this candidate based on received votes.
+	approval_stake: Balance,
+	// Intermediary value used to sort candidates.
+	// See Phragmén reference implementation.
+	pub score: Perquintill,
+}
+
+// Wrapper around the nomination info of a single nominator for a group of validators.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Nominations<AccountId, Balance: HasCompact> {
+	// The nominator's account.
+	who: AccountId,
+	// List of validators proposed by this nominator.
+	nominees: Vec<Vote<AccountId, Balance>>,
+	// the stake amount proposed by the nominator as a part of the vote.
+	// Same as `nom.budget` in Phragmén reference.
+	stake: Balance,
+	// Incremented each time a nominee that this nominator voted for has been elected.
+	load: Perquintill,
+}
+
+// Wrapper around a nominator vote and the load of that vote.
+// Referred to as 'edge' in the Phragmén reference implementation.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Vote<AccountId, Balance: HasCompact> {
+	// Account being voted for
+	who: AccountId,
+	// Load of this vote.
+	load: Perquintill,
+	// Final backing stake of this vote.
+	backing_stake: Balance
+}
+
+/// Perform election based on Phragmén algorithm.
+///
+/// Reference implementation: https://github.com/w3f/consensus
+///
+/// @returns a vector of elected candidates
+pub fn elect<T: Trait + 'static, FR, FN, FV, FS>(
+		get_rounds: FR,
+		get_validators: FV,
+		get_nominators: FN,
+		stash_of: FS,
+		minimum_validator_count: usize,
+	) -> Vec<Candidate<T::AccountId, BalanceOf<T>>> where
+		FR: Fn() -> usize,
+		FV: Fn() -> Box<dyn Iterator<
```

Comment:
`Box<..>` looks a bit nasty..

Reply:
This is actually the type returned by the `.enumerate()`; passing in the storage item as it via a closure led to an even more nasty type. 

What looks cleaner but was not applied in favor of keeping the code in `lib.rs` as simple as possible is to pre-process everything in `lib.rs` and pass them in as a closure that returns e.g. a Vector of validators. Though, this is computationaly not optimal in my opinion.

Reply:
ok - something to check for optimisation at a later date.